src/interface/state_descriptor.hpp

```diff
@@ -110,6 +111,60 @@ class StateDescriptor {
     params_.Add<T>(key, value, is_mutable);
   }
 
+  // JMM: The additional arguments here are to support additional
+  // docstring arguments for ParameterInput down the line
+  // I wish I could figure out how to do this with fewer overloads...
+  template <typename T, typename... Args>
+  void AddParamFromInput(Params::Mutability mutability, const std::string &block,
+                         const std::string &name, ParameterInput *pinput,
+                         Args &&...args) {
+    T val = pinput->template Get<T>(block, name, std::forward<Args>(args)...);
+    params_.Add(name, val, mutability);
+  }
+  template <typename T, typename... Args>
+  void AddParamFromInput(Params::Mutability mutability, const std::string &block,
+                         const std::string &name, const T &default_val,
+                         ParameterInput *pinput, Args &&...args) {
+    T val = pinput->template GetOrAdd<T>(block, name, default_val,
+                                         std::forward<Args>(args)...);
+    params_.Add(name, val, mutability);
+  }
```

Comment:
I added this set of functions in this MR but they're not really relevant for the bulk of the changes. These are just syntactic sugar.